examples/src/big_compute_buffers/mod.rs

```diff
@@ -0,0 +1,288 @@
+const MAX_BUFFER_SIZE: u64 = 1 << 27; // 134_217_728 // 134MB
+const MAX_DISPATCH_SIZE: u32 = (1 << 16) - 1;
+
+pub async fn execute_gpu(numbers: &[f32]) -> Option<Vec<f32>> {
+    let instance = wgpu::Instance::default();
+
+    let adapter = instance
+        .request_adapter(&wgpu::RequestAdapterOptions::default())
+        .await?;
+
+    let (device, queue) = adapter
+        .request_device(
+            &wgpu::DeviceDescriptor {
+                label: None,
+                // These features are required to use `binding_array` in your wgsl.
+                // Without them your shader may fail to compile.
+                required_features: Features::STORAGE_RESOURCE_BINDING_ARRAY
+                    | Features::BUFFER_BINDING_ARRAY
+                    | Features::UNIFORM_BUFFER_AND_STORAGE_TEXTURE_ARRAY_NON_UNIFORM_INDEXING,
+
+                memory_hints: wgpu::MemoryHints::Performance,
+                required_limits: wgpu::Limits {
+                    max_buffer_size: MAX_BUFFER_SIZE,
+                    ..Default::default()
+                },
+            },
+            None,
+        )
+        .await
+        .unwrap();
+
+    execute_gpu_inner(&device, &queue, numbers).await
+}
+
+pub async fn execute_gpu_inner(
+    device: &wgpu::Device,
+    queue: &wgpu::Queue,
+    numbers: &[f32],
+) -> Option<Vec<f32>> {
+    let (staging_buffers, storage_buffers, bind_group, compute_pipeline) = setup(device, numbers);
+
+    let mut encoder =
+        device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
+    {
+        let mut cpass = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
+            label: Some("compute pass descriptor"),
+            timestamp_writes: None,
+        });
+        cpass.set_pipeline(&compute_pipeline);
+        cpass.set_bind_group(0, Some(&bind_group), &[]);
+
+        cpass.dispatch_workgroups(MAX_DISPATCH_SIZE.min(numbers.len() as u32), 1, 1);
+    }
+
+    storage_buffers.iter().zip(staging_buffers.iter()).for_each(
+        |(storage_buffer, staging_buffer)| {
+            let stg_size = staging_buffer.size();
+
+            encoder.copy_buffer_to_buffer(
+                storage_buffer, // Source buffer
+                0,
+                staging_buffer, // Destination buffer
+                0,
+                stg_size,
+            );
+        },
+    );
+
+    queue.submit(Some(encoder.finish()));
+
+    let mut buffer_slices = Vec::new();
+    staging_buffers.iter().for_each(|sb| {
+        buffer_slices.push(sb.slice(..));
+    });
+
+    let (sender, receiver) = flume::bounded(buffer_slices.len());
+    let sender = Arc::new(sender);
+
+    buffer_slices.iter().for_each(|bs| {
+        let sender = sender.clone();
+        bs.map_async(wgpu::MapMode::Read, move |v| {
+            sender.send(v).unwrap();
+        })
+    });
+
+    device.poll(wgpu::Maintain::wait());
+
+    if let Ok(Ok(())) = receiver.recv_async().await {
```

Comment:
It is more legible to handle errors immediately, and let the function's main flow of control (the least indented) be the success path. That is:

```
let Ok(Ok(())) = receiver.recv_async().await else {
    log::error!("Failed to run compute on GPU!");
    return None;
};
```

and then have the success case, where we iterate over `buffer_slices`, follow directly.